False negative in LeftCurly for types

I have read check documentation: https://checkstyle.sourceforge.io/checks/blocks/leftcurly.html#LeftCurly
I have read https://checkstyle.sourceforge.io/property_types.html#LeftCurlyOption
I have downloaded the latest checkstyle from: https://checkstyle.org/cmdline.html#Download_and_Run
I have executed the cli and showed it below, as cli describes the problem better than 1,000 words

Using confg:
```
atharv@atharv-IdeaPad-3-15IIL05:~/test$ cat config.xml
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
  "-//Puppy Crawl//DTD Check Configuration 1.3//EN"
  "https://checkstyle.org/dtds/configuration_1_3.dtd">
<module name="Checker">
  <module name="TreeWalker">
    <module name="LeftCurly">
      <property name="option" value="eol"/>
    </module>
  </module>
</module>
```

Types
```java
atharv@atharv-IdeaPad-3-15IIL05:~/test$ cat Test.java
public class Test { int a = 1;      // expected violation
}

interface Test1 { int b = 1;        // expected violation
}

@interface Test2 { int c = 1;       // expected violation
}

record Test3() { static int k = 0;  // expected violation
}
```
```
atharv@atharv-IdeaPad-3-15IIL05:~/test$ java -jar checkstyle-14.1.0-all.jar -c config.xml Test.java
Starting audit...
Audit done.
```
Expected:
```
Starting audit...
[ERROR] /home/atharv/test/Test.java:1:19: '{' at column 19 should have line break after. [LeftCurly]
[ERROR] /home/atharv/test/Test.java:4:17: '{' at column 17 should have line break after. [LeftCurly]
[ERROR] /home/atharv/test/Test.java:7:18: '{' at column 18 should have line break after. [LeftCurly]
[ERROR] /home/atharv/test/Test.java:10:16: '{' at column 16 should have line break after. [LeftCurly]
Audit done.
Checkstyle ends with 4 errors.
```
---

Switch
```java
atharv@atharv-IdeaPad-3-15IIL05:~/test$ cat Test.java
public class Test {
    void method() {
        int a = 1;
        switch (a) {default: a+= 1;} // expected violation
    }
}
```
```
atharv@atharv-IdeaPad-3-15IIL05:~/test$ java -jar checkstyle-14.1.0-all.jar -c config.xml Test.java
Starting audit...
Audit done.
```
Expected:
```
Starting audit...
[ERROR] /home/atharv/test/Test.java:4:20: '{' at column 20 should have line break after. [LeftCurly]
Audit done.
Checkstyle ends with 1 errors.
```
---

Anonymous Class
```java
atharv@atharv-IdeaPad-3-15IIL05:~/test$ cat Test.java
public class Test {

    void method() {
        Runnable runnable = new Runnable() { @Override
            public void run() {
            }
        };
    }
}
```
```
atharv@atharv-IdeaPad-3-15IIL05:~/test$ java -jar checkstyle-14.1.0-all.jar -c config.xml Test.java
Starting audit...
Audit done.
```
Expected:
```
Starting audit...
[ERROR] /home/atharv/test/Test.java:4:45: '{' at column 45 should have line break after. [LeftCurly]
Audit done.
Checkstyle ends with 1 errors.
```

Anonymous Class violation is expected because, violation is reported for option [nl](https://checkstyle.sourceforge.io/property-types.html#LeftCurlyOption-nl)

With config;
```
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
  "-//Puppy Crawl//DTD Check Configuration 1.3//EN"
  "https://checkstyle.org/dtds/configuration_1_3.dtd">
<module name="Checker">
  <module name="TreeWalker">
    <module name="LeftCurly">
      <property name="option" value="nl"/>
    </module>
  </module>
</module>
```
```java
atharv@atharv-IdeaPad-3-15IIL05:~/test$ cat Test.java
public class Test 
{

    void method() 
    {
        Runnable runnable = new Runnable() {
            @Override
            public void run() 
            {
            }
        };
    }
}
```
```
atharv@atharv-IdeaPad-3-15IIL05:~/test$ java -jar checkstyle-14.1.0-all.jar -c config.xml Test.java
Starting audit...
[ERROR] /home/atharv/test/Test.java:6:44: '{' at column 44 should be on a new line. [LeftCurly]
Audit done.
Checkstyle ends with 1 errors
```

--- a/config/google-java-format/excluded/compilable-input-paths.txt
+++ b/config/google-java-format/excluded/compilable-input-paths.txt
@@ -11,6 +11,7 @@ src/it/resources/com/google/checkstyle/test/chapter3filestructure/rule333orderin
 src/it/resources/com/google/checkstyle/test/chapter4formatting/rule412nonemptyblocks/InputNonemptyBlocksLeftRightCurly.java
 src/it/resources/com/google/checkstyle/test/chapter4formatting/rule412nonemptyblocks/InputLeftCurlyAnnotations.java
 src/it/resources/com/google/checkstyle/test/chapter4formatting/rule412nonemptyblocks/InputLeftCurlyMethod.java
+src/it/resources/com/google/checkstyle/test/chapter4formatting/rule412nonemptyblocks/InputLeftCurlyTypesAndAnonymousClass.java
 src/it/resources/com/google/checkstyle/test/chapter4formatting/rule412nonemptyblocks/InputRightCurly.java
 src/it/resources/com/google/checkstyle/test/chapter4formatting/rule412nonemptyblocks/InputRightCurlyOther.java
 src/it/resources/com/google/checkstyle/test/chapter4formatting/rule412nonemptyblocks/InputRightCurlySwitchCase.java

--- a/src/it/java/com/google/checkstyle/test/chapter4formatting/rule412nonemptyblocks/NonemptyBlocksKrStyleTest.java
+++ b/src/it/java/com/google/checkstyle/test/chapter4formatting/rule412nonemptyblocks/NonemptyBlocksKrStyleTest.java
@@ -140,4 +140,14 @@ public class NonemptyBlocksKrStyleTest extends AbstractGoogleModuleTestSupport {
         verifyWithWholeConfig(getPath("InputFormattedTryCatchIfElse2.java"));
     }
 
+    @Test
+    public void testLeftCurlyTypesAndAnonymousClass() throws Exception {
+        verifyWithWholeConfig(getPath("InputLeftCurlyTypesAndAnonymousClass.java"));
+    }
+
+    @Test
+    public void testLeftCurlyTypesAndAnonymousClassFormatted() throws Exception {
+        verifyWithWholeConfig(getPath("InputFormattedLeftCurlyTypesAndAnonymousClass.java"));
+    }
+
 }

--- a/src/it/resources/com/google/checkstyle/test/chapter4formatting/rule412nonemptyblocks/InputFormattedLeftCurlyTypesAndAnonymousClass.java
+++ b/src/it/resources/com/google/checkstyle/test/chapter4formatting/rule412nonemptyblocks/InputFormattedLeftCurlyTypesAndAnonymousClass.java
@@ -1,0 +1,57 @@
+package com.google.checkstyle.test.chapter4formatting.rule412nonemptyblocks;
+
+/** Some javadoc. */
+public class InputFormattedLeftCurlyTypesAndAnonymousClass {
+
+  /** Some javadoc. */
+  public void method() {
+    final Runnable r1 =
+        new Runnable() {
+          @Override
+          public void run() {}
+        };
+
+    final Thread t =
+        new Thread() {
+          @Override
+          public void run() {}
+        };
+  }
+
+  /** Some javadoc. */
+  public void anotherMethod() {
+    final Runnable runnable =
+        new Runnable() {
+          @Override
+          public void run() {}
+        };
+
+    final Comparable<Integer> comparable =
+        new Comparable<Integer>() {
+          @Override
+          public int compareTo(Integer o) {
+            return 0;
+          }
+        };
+
+    final Object obj = new Object() {};
+  }
+
+  static class InnerA {
+    int value = 1;
+    int value2 = 2;
+  }
+
+  record InnerB() {
+    static int num = 1;
+  }
+
+  interface InnerC {
+    void doSomething();
+  }
+
+  static class InnerD extends Thread {
+    @Override
+    public void run() {}
+  }
+}

--- a/src/it/resources/com/google/checkstyle/test/chapter4formatting/rule412nonemptyblocks/InputLeftCurlyTypesAndAnonymousClass.java
+++ b/src/it/resources/com/google/checkstyle/test/chapter4formatting/rule412nonemptyblocks/InputLeftCurlyTypesAndAnonymousClass.java
@@ -1,0 +1,58 @@
+package com.google.checkstyle.test.chapter4formatting.rule412nonemptyblocks;
+
+/** Some javadoc. */
+public class InputLeftCurlyTypesAndAnonymousClass {
+
+  /** Some javadoc. */
+  public void method() {
+    final Runnable r1 =
+        new Runnable() {
+          @Override
+          public void run() {}
+        };
+
+    final Thread t = new Thread()
+      {
+        // violation above ''{' at column 7 should be on the previous line'
+        @Override
+        public void run() {}
+      };
+  }
+
+  /** Some javadoc. */
+  public void anotherMethod() {
+    // violation 2 lines below ''{' at column 24 should have line break after'
+    final Runnable runnable =
+        new Runnable() { @Override
+          public void run() {}
+        };
+
+    final Comparable<Integer> comparable =
+        new Comparable<Integer>() {
+          @Override
+          public int compareTo(Integer o) {
+            return 0;
+          }
+        };
+
+    final Object obj = new Object() {};
+  }
+
+  // violation below ''{' at column 23 should have line break after'
+  static class InnerA { int value = 1;
+    int value2 = 2;
+  }
+
+  // violation below ''{' at column 19 should have line break after'
+  record InnerB() { static
+      int num = 1;
+  }
+
+  // violation below ''{' at column 20 should have line break after'
+  interface InnerC { void doSomething(); }
+
+  static class InnerD extends Thread {
+    @Override
+    public void run() {}
+  }
+}

--- a/src/it/resources/com/google/checkstyle/test/chapter4formatting/rule43onestatement/InputOneStatementPerLineEdgeCases.java
+++ b/src/it/resources/com/google/checkstyle/test/chapter4formatting/rule43onestatement/InputOneStatementPerLineEdgeCases.java
@@ -36,7 +36,8 @@ public class InputOneStatementPerLineEdgeCases {
 
     Object outerA = new Object() {
         Object innerA = new Object() { void foo() { int a; int b; } }; };
-    // 4 violations above:
+    // 5 violations above:
+    // ''{' at column 38 should have line break after.'
     // ''{' at column 51 should have line break after.'
     // 'Only one variable definition per line allowed.'
     // 'Only one statement per line allowed.'

--- a/src/it/resources/com/google/checkstyle/test/chapter4formatting/rule4841indentation/InputLambdaChild.java
+++ b/src/it/resources/com/google/checkstyle/test/chapter4formatting/rule4841indentation/InputLambdaChild.java
@@ -29,11 +29,14 @@ public class InputLambdaChild {
         .findFirst().orElse("defaultValue");
   }
 
+  // violation 3 lines below ''{' at column 56 should have line break after'
+  // violation 3 lines below ''{' at column 56 should have line break after'
   void main(String[] args) {
     group((Function<Integer, Integer>) x -> switch (x) { default: yield x; },
           (Function<Integer, Integer>) x -> switch (x) { default: yield x; });
   }
 
+  // violation 3 lines below ''{' at column 31 should have line break after'
   List<String> getThrowsTrees(Object input) {
     return getBlockTags(input,
         kind -> switch (kind) { case "EXCEPTION", "THROWS" -> true; default -> false; },

--- a/src/main/java/com/puppycrawl/tools/checkstyle/checks/blocks/LeftCurlyCheck.java
+++ b/src/main/java/com/puppycrawl/tools/checkstyle/checks/blocks/LeftCurlyCheck.java
@@ -21,7 +21,6 @@ package com.puppycrawl.tools.checkstyle.checks.blocks;
 
 import java.util.Locale;
 import java.util.Optional;
-import java.util.Set;
 
 import javax.annotation.Nullable;
 
@@ -30,6 +29,7 @@ import com.puppycrawl.tools.checkstyle.api.AbstractCheck;
 import com.puppycrawl.tools.checkstyle.api.DetailAST;
 import com.puppycrawl.tools.checkstyle.api.TokenTypes;
 import com.puppycrawl.tools.checkstyle.utils.CommonUtil;
+import com.puppycrawl.tools.checkstyle.utils.NullUtil;
 import com.puppycrawl.tools.checkstyle.utils.TokenUtil;
 
 /**
@@ -179,53 +179,24 @@ public class LeftCurlyCheck
                 startToken = ast;
                 yield getBraceFromSwitchMember(ast);
             }
+            case TokenTypes.OBJBLOCK -> {
+                startToken = ast;
+                DetailAST braceToken = null;
+                if (ast.getParent().getType() == TokenTypes.LITERAL_NEW) {
+                    braceToken = ast;
+                }
+                yield braceToken;
+            }
             default -> {
-                // ATTENTION! We have default here, but we expect case TokenTypes.METHOD_DEF,
-                // TokenTypes.LITERAL_FOR, TokenTypes.LITERAL_WHILE, TokenTypes.LITERAL_DO only.
-                // It has been done to improve coverage to 100%. I couldn't replace it with
-                // if-else-if block because code was ugly and didn't pass pmd check.
-
+                // only expected DEFAULT Token is LITERAL_SWITCH
                 startToken = ast;
                 yield ast.findFirstToken(TokenTypes.LCURLY);
             }
         };
 
-        if (brace != null && !isBraceVerifiedByParent(ast)) {
+        if (brace != null) {
             verifyBrace(brace, startToken);
         }
-    }
-
-    /**
-     * Checks whether the brace of the given token is verified through its parent
-     * as well. An {@code OBJBLOCK} shares its brace with the type definition it
-     * belongs to, and the case label of an arrow switch shares its brace with the
-     * {@code SWITCH_RULE}. Without this the same brace is reported twice.
-     *
-     * @param ast the token being visited
-     * @return {@code true} if the parent already verifies the same brace
-     */
-    private boolean isBraceVerifiedByParent(DetailAST ast) {
-        return TokenUtil.isOfType(ast,
-                    TokenTypes.OBJBLOCK, TokenTypes.LITERAL_CASE, TokenTypes.LITERAL_DEFAULT)
-                && isConfigured(ast.getParent());
-    }
-
-    /**
-     * Checks whether the token type of the given node is configured for this check.
-     *
-     * @param ast the token to check
-     * @return {@code true} if this check visits that token type
-     */
-    private boolean isConfigured(DetailAST ast) {
-        final Set<String> configuredTokens = getTokenNames();
-        final boolean result;
-        if (configuredTokens.isEmpty()) {
-            result = TokenUtil.isOfType(ast, getDefaultTokens());
-        }
-        else {
-            result = configuredTokens.contains(TokenUtil.getTokenName(ast.getType()));
-        }
-        return result;
     }
 
     /**
@@ -330,7 +301,7 @@ public class LeftCurlyCheck
                 }
             }
             else if (option == LeftCurlyOption.EOL) {
-                validateEol(brace, braceLine);
+                validateEol(startToken, brace, braceLine);
             }
             else if (!TokenUtil.areOnSameLine(startToken, brace)) {
                 validateNewLinePosition(brace, startToken, braceLine);
@@ -341,14 +312,15 @@ public class LeftCurlyCheck
     /**
      * Validate EOL case.
      *
+     * @param startToken token for start of expression.
      * @param brace brace AST
      * @param braceLine line content
      */
-    private void validateEol(DetailAST brace, String braceLine) {
+    private void validateEol(DetailAST startToken, DetailAST brace, String braceLine) {
         if (CommonUtil.hasWhitespaceBefore(brace.getColumnNo(), braceLine)) {
             log(brace, MSG_KEY_LINE_PREVIOUS, OPEN_CURLY_BRACE, brace.getColumnNo() + 1);
         }
-        if (!hasLineBreakAfter(brace)) {
+        if (!hasLineBreakAfter(startToken, brace)) {
             log(brace, MSG_KEY_LINE_BREAK_AFTER, OPEN_CURLY_BRACE, brace.getColumnNo() + 1);
         }
     }
@@ -378,28 +350,30 @@ public class LeftCurlyCheck
     /**
      * Checks if left curly has line break after.
      *
-     * @param leftCurly
+     * @param startToken token for start of expression.
+     * @param curlyBrace
      *        Left curly token.
      * @return
      *        True, left curly has line break after.
      */
-    private boolean hasLineBreakAfter(DetailAST leftCurly) {
-        DetailAST nextToken = null;
-        if (leftCurly.getType() == TokenTypes.SLIST) {
-            nextToken = leftCurly.getFirstChild();
+    private boolean hasLineBreakAfter(DetailAST startToken, DetailAST curlyBrace) {
+        DetailAST nextToken = curlyBrace.getNextSibling();
+        if (curlyBrace.getType() == TokenTypes.OBJBLOCK
+                && (!ignoreEnums || startToken.getType() != TokenTypes.ENUM_DEF)) {
+            nextToken = NullUtil.notNull(curlyBrace.findFirstToken(TokenTypes.LCURLY))
+                    .getNextSibling();
+
         }
-        else if (!ignoreEnums) {
-            if (leftCurly.getParent().getParent().getType() == TokenTypes.ENUM_DEF) {
-                nextToken = leftCurly.getNextSibling();
-            }
-            else if (leftCurly.getParent().getType() == TokenTypes.ENUM_DEF) {
-                // the brace of the enum body is the first child of its OBJBLOCK
-                nextToken = leftCurly.getFirstChild().getNextSibling();
-            }
+        else if (curlyBrace.getType() == TokenTypes.SLIST) {
+            nextToken = curlyBrace.getFirstChild();
+        }
+        if (nextToken != null && nextToken.getType() == TokenTypes.INSTANCE_INIT
+                && startToken.getType() == TokenTypes.OBJBLOCK) {
+            nextToken = null;
         }
         return nextToken == null
                 || nextToken.getType() == TokenTypes.RCURLY
-                || !TokenUtil.areOnSameLine(leftCurly, nextToken);
+                || !TokenUtil.areOnSameLine(curlyBrace, nextToken);
     }
 
 }

--- a/src/main/java/com/puppycrawl/tools/checkstyle/checks/blocks/LeftCurlyCheck.java
+++ b/src/main/java/com/puppycrawl/tools/checkstyle/checks/blocks/LeftCurlyCheck.java
@@ -29,6 +29,7 @@ import com.puppycrawl.tools.checkstyle.api.AbstractCheck;
 import com.puppycrawl.tools.checkstyle.api.DetailAST;
 import com.puppycrawl.tools.checkstyle.api.TokenTypes;
 import com.puppycrawl.tools.checkstyle.utils.CommonUtil;
+import com.puppycrawl.tools.checkstyle.utils.NullUtil;
 import com.puppycrawl.tools.checkstyle.utils.TokenUtil;
 
 /**
@@ -178,12 +179,16 @@ public class LeftCurlyCheck
                 startToken = ast;
                 yield getBraceFromSwitchMember(ast);
             }
+            case TokenTypes.OBJBLOCK -> {
+                startToken = ast;
+                DetailAST braceToken = null;
+                if (ast.getParent().getType() == TokenTypes.LITERAL_NEW) {
+                    braceToken = ast;
+                }
+                yield braceToken;
+            }
             default -> {
-                // ATTENTION! We have default here, but we expect case TokenTypes.METHOD_DEF,
-                // TokenTypes.LITERAL_FOR, TokenTypes.LITERAL_WHILE, TokenTypes.LITERAL_DO only.
-                // It has been done to improve coverage to 100%. I couldn't replace it with
-                // if-else-if block because code was ugly and didn't pass pmd check.
-
+                // ATTENTION! We have default here, but we expect case TokenTypes.LITERAL_SWITCH
                 startToken = ast;
                 yield ast.findFirstToken(TokenTypes.LCURLY);
             }
@@ -296,7 +301,7 @@ public class LeftCurlyCheck
                 }
             }
             else if (option == LeftCurlyOption.EOL) {
-                validateEol(brace, braceLine);
+                validateEol(startToken, brace, braceLine);
             }
             else if (!TokenUtil.areOnSameLine(startToken, brace)) {
                 validateNewLinePosition(brace, startToken, braceLine);
@@ -307,14 +312,15 @@ public class LeftCurlyCheck
     /**
      * Validate EOL case.
      *
+     * @param startToken token for start of expression.
      * @param brace brace AST
      * @param braceLine line content
      */
-    private void validateEol(DetailAST brace, String braceLine) {
+    private void validateEol(DetailAST startToken, DetailAST brace, String braceLine) {
         if (CommonUtil.hasWhitespaceBefore(brace.getColumnNo(), braceLine)) {
             log(brace, MSG_KEY_LINE_PREVIOUS, OPEN_CURLY_BRACE, brace.getColumnNo() + 1);
         }
-        if (!hasLineBreakAfter(brace)) {
+        if (!hasLineBreakAfter(startToken, brace)) {
             log(brace, MSG_KEY_LINE_BREAK_AFTER, OPEN_CURLY_BRACE, brace.getColumnNo() + 1);
         }
     }
@@ -344,21 +350,26 @@ public class LeftCurlyCheck
     /**
      * Checks if left curly has line break after.
      *
+     * @param startToken token for start of expression.
      * @param leftCurly
      *        Left curly token.
      * @return
      *        True, left curly has line break after.
      */
-    private boolean hasLineBreakAfter(DetailAST leftCurly) {
-        DetailAST nextToken = null;
-        if (leftCurly.getType() == TokenTypes.SLIST) {
+    private boolean hasLineBreakAfter(DetailAST startToken, DetailAST leftCurly) {
+        DetailAST nextToken = leftCurly.getNextSibling();
+        if (leftCurly.getType() == TokenTypes.OBJBLOCK
+                && (!ignoreEnums || startToken.getType() != TokenTypes.ENUM_DEF)) {
+            nextToken = NullUtil.notNull(leftCurly.findFirstToken(TokenTypes.LCURLY))
+                    .getNextSibling();
+
+        }
+        else if (leftCurly.getType() == TokenTypes.SLIST) {
             nextToken = leftCurly.getFirstChild();
         }
-        else {
-            if (!ignoreEnums
-                    && leftCurly.getParent().getParent().getType() == TokenTypes.ENUM_DEF) {
-                nextToken = leftCurly.getNextSibling();
-            }
+        if (nextToken != null && nextToken.getType() == TokenTypes.INSTANCE_INIT
+                && startToken.getType() == TokenTypes.OBJBLOCK) {
+            nextToken = null;
         }
         return nextToken == null
                 || nextToken.getType() == TokenTypes.RCURLY

--- a/src/test/java/com/puppycrawl/tools/checkstyle/checks/blocks/LeftCurlyCheckTest.java
+++ b/src/test/java/com/puppycrawl/tools/checkstyle/checks/blocks/LeftCurlyCheckTest.java
@@ -365,6 +365,7 @@ public class LeftCurlyCheckTest extends AbstractModuleTestSupport {
     public void testIgnoreEnumsOptionTrue() throws Exception {
         final String[] expectedWhileTrue = {
             "21:44: " + getCheckMessage(MSG_KEY_LINE_BREAK_AFTER, "{", 44),
+            "34:20: " + getCheckMessage(MSG_KEY_LINE_BREAK_AFTER, "{", 20),
         };
         verifyWithInlineConfigParser(
                 getPath("InputLeftCurlyIgnoreEnumsOptTrue.java"), expectedWhileTrue);
@@ -375,6 +376,7 @@ public class LeftCurlyCheckTest extends AbstractModuleTestSupport {
         final String[] expectedWhileFalse = {
             "17:17: " + getCheckMessage(MSG_KEY_LINE_BREAK_AFTER, "{", 17),
             "21:44: " + getCheckMessage(MSG_KEY_LINE_BREAK_AFTER, "{", 44),
+            "34:20: " + getCheckMessage(MSG_KEY_LINE_BREAK_AFTER, "{", 20),
         };
         verifyWithInlineConfigParser(
                 getPath("InputLeftCurlyIgnoreEnumsOptFalse.java"), expectedWhileFalse);
@@ -663,6 +665,30 @@ public class LeftCurlyCheckTest extends AbstractModuleTestSupport {
         };
         verifyWithInlineConfigParser(
                 getPath("InputLeftCurlySwitchMutation.java"), expected);
+    }
+
+    @Test
+    public void testTypesEol() throws Exception {
+        final String[] expected = {
+            "19:11: " + getCheckMessage(MSG_KEY_LINE_BREAK_AFTER, "{", 11),
+            "23:15: " + getCheckMessage(MSG_KEY_LINE_BREAK_AFTER, "{", 15),
+            "27:17: " + getCheckMessage(MSG_KEY_LINE_BREAK_AFTER, "{", 17),
+            "31:12: " + getCheckMessage(MSG_KEY_LINE_BREAK_AFTER, "{", 12),
+            "36:17: " + getCheckMessage(MSG_KEY_LINE_BREAK_AFTER, "{", 17),
+            "40:12: " + getCheckMessage(MSG_KEY_LINE_BREAK_AFTER, "{", 12),
+        };
+        verifyWithInlineConfigParser(
+                getPath("InputLeftCurlyTestTypesEol.java"), expected);
+    }
+
+    @Test
+    public void testAnonymousClassEol() throws Exception {
+        final String[] expected = {
+            "28:38: " + getCheckMessage(MSG_KEY_LINE_BREAK_AFTER, "{", 38),
+            "36:9: " + getCheckMessage(MSG_KEY_LINE_PREVIOUS, "{", 9),
+        };
+        verifyWithInlineConfigParser(
+                getPath("InputLeftCurlyAnonymousClassEol.java"), expected);
     }
 
 }

--- a/src/test/java/com/puppycrawl/tools/checkstyle/checks/blocks/LeftCurlyCheckTest.java
+++ b/src/test/java/com/puppycrawl/tools/checkstyle/checks/blocks/LeftCurlyCheckTest.java
@@ -365,6 +365,7 @@ public class LeftCurlyCheckTest extends AbstractModuleTestSupport {
     public void testIgnoreEnumsOptionTrue() throws Exception {
         final String[] expectedWhileTrue = {
             "21:44: " + getCheckMessage(MSG_KEY_LINE_BREAK_AFTER, "{", 44),
+            "34:20: " + getCheckMessage(MSG_KEY_LINE_BREAK_AFTER, "{", 20),
         };
         verifyWithInlineConfigParser(
                 getPath("InputLeftCurlyIgnoreEnumsOptTrue.java"), expectedWhileTrue);
@@ -375,6 +376,7 @@ public class LeftCurlyCheckTest extends AbstractModuleTestSupport {
         final String[] expectedWhileFalse = {
             "17:17: " + getCheckMessage(MSG_KEY_LINE_BREAK_AFTER, "{", 17),
             "21:44: " + getCheckMessage(MSG_KEY_LINE_BREAK_AFTER, "{", 44),
+            "34:20: " + getCheckMessage(MSG_KEY_LINE_BREAK_AFTER, "{", 20),
         };
         verifyWithInlineConfigParser(
                 getPath("InputLeftCurlyIgnoreEnumsOptFalse.java"), expectedWhileFalse);
@@ -383,7 +385,7 @@ public class LeftCurlyCheckTest extends AbstractModuleTestSupport {
     @Test
     public void testObjBlockTokenOnly() throws Exception {
         final String[] expected = {
-            "13:17: " + getCheckMessage(MSG_KEY_LINE_BREAK_AFTER, "{", 17),
+            "24:41: " + getCheckMessage(MSG_KEY_LINE_BREAK_AFTER, "{", 41),
         };
         verifyWithInlineConfigParser(
                 getPath("InputLeftCurlyObjBlockTokenOnly.java"), expected);
@@ -696,6 +698,30 @@ public class LeftCurlyCheckTest extends AbstractModuleTestSupport {
         };
         verifyWithInlineConfigParser(
                 getPath("InputLeftCurlySwitchWhen.java"), expected);
+    }
+
+    @Test
+    public void testTypesEol() throws Exception {
+        final String[] expected = {
+            "19:11: " + getCheckMessage(MSG_KEY_LINE_BREAK_AFTER, "{", 11),
+            "23:15: " + getCheckMessage(MSG_KEY_LINE_BREAK_AFTER, "{", 15),
+            "27:17: " + getCheckMessage(MSG_KEY_LINE_BREAK_AFTER, "{", 17),
+            "31:12: " + getCheckMessage(MSG_KEY_LINE_BREAK_AFTER, "{", 12),
+            "36:17: " + getCheckMessage(MSG_KEY_LINE_BREAK_AFTER, "{", 17),
+            "40:12: " + getCheckMessage(MSG_KEY_LINE_BREAK_AFTER, "{", 12),
+        };
+        verifyWithInlineConfigParser(
+                getPath("InputLeftCurlyTestTypesEol.java"), expected);
+    }
+
+    @Test
+    public void testAnonymousClassEol() throws Exception {
+        final String[] expected = {
+            "28:38: " + getCheckMessage(MSG_KEY_LINE_BREAK_AFTER, "{", 38),
+            "36:9: " + getCheckMessage(MSG_KEY_LINE_PREVIOUS, "{", 9),
+        };
+        verifyWithInlineConfigParser(
+                getPath("InputLeftCurlyAnonymousClassEol.java"), expected);
     }
 
 }

--- a/src/test/resources/com/puppycrawl/tools/checkstyle/checks/blocks/leftcurly/InputLeftCurlyAnonymousClassEol.java
+++ b/src/test/resources/com/puppycrawl/tools/checkstyle/checks/blocks/leftcurly/InputLeftCurlyAnonymousClassEol.java
@@ -1,0 +1,50 @@
+/*
+LeftCurly
+option = (default)eol
+ignoreEnums = (default)true
+tokens = (default)ANNOTATION_DEF, CLASS_DEF, CTOR_DEF, ENUM_CONSTANT_DEF, ENUM_DEF, \
+         INTERFACE_DEF, LAMBDA, LITERAL_CASE, LITERAL_CATCH, LITERAL_DEFAULT, \
+         LITERAL_DO, LITERAL_ELSE, LITERAL_FINALLY, LITERAL_FOR, LITERAL_IF, \
+         LITERAL_SWITCH, LITERAL_SYNCHRONIZED, LITERAL_TRY, LITERAL_WHILE, METHOD_DEF, \
+         OBJBLOCK, STATIC_INIT, RECORD_DEF, COMPACT_CTOR_DEF, SWITCH_RULE
+
+
+*/
+
+package com.puppycrawl.tools.checkstyle.checks.blocks.leftcurly;
+
+import java.util.HashMap;
+import java.util.Map;
+
+public class InputLeftCurlyAnonymousClassEol {
+
+    Map<Integer, Integer> map = new HashMap<>() {{
+        map.put(1, 2);
+        map.put(1, 3);
+    }};
+
+    void method() {
+        // violation below ''{' at column 38 should have line break after'
+        Runnable r1 = new Runnable() { @Override
+            public void run() {
+
+            }
+        };
+
+        // violation 2 lines below ''{' at column 9 should be on the previous line'
+        Runnable r2 = new Runnable()
+        {
+            @Override
+            public void run() {
+
+            }
+        };
+
+        Thread t = new Thread() {
+            @Override
+            public void run() {
+
+            }
+        };
+    }
+}

--- a/src/test/resources/com/puppycrawl/tools/checkstyle/checks/blocks/leftcurly/InputLeftCurlyIgnoreEnumsOptFalse.java
+++ b/src/test/resources/com/puppycrawl/tools/checkstyle/checks/blocks/leftcurly/InputLeftCurlyIgnoreEnumsOptFalse.java
@@ -30,6 +30,7 @@ public class InputLeftCurlyIgnoreEnumsOptFalse {
     }
 
     void method1(int a) {
+        // violation below ''{' at column 20 should have line break after'
         switch (a) {case 1: ; }
     }
 }

--- a/src/test/resources/com/puppycrawl/tools/checkstyle/checks/blocks/leftcurly/InputLeftCurlyIgnoreEnumsOptTrue.java
+++ b/src/test/resources/com/puppycrawl/tools/checkstyle/checks/blocks/leftcurly/InputLeftCurlyIgnoreEnumsOptTrue.java
@@ -30,6 +30,7 @@ public class InputLeftCurlyIgnoreEnumsOptTrue {
     }
 
     void method1(int a) {
+        // violation below ''{' at column 20 should have line break after'
         switch (a) {case 1: ; }
     }
 }

--- a/src/test/resources/com/puppycrawl/tools/checkstyle/checks/blocks/leftcurly/InputLeftCurlyObjBlockTokenOnly.java
+++ b/src/test/resources/com/puppycrawl/tools/checkstyle/checks/blocks/leftcurly/InputLeftCurlyObjBlockTokenOnly.java
@@ -10,7 +10,7 @@ tokens = OBJBLOCK
 package com.puppycrawl.tools.checkstyle.checks.blocks.leftcurly;
 
 public class InputLeftCurlyObjBlockTokenOnly {
-    enum Colors {RED, // violation ''{' at column 17 should have line break after'
+    enum Colors {RED,
         BLUE,
         GREEN;
     }
@@ -20,8 +20,8 @@ public class InputLeftCurlyObjBlockTokenOnly {
         CIRCLE;
     }
 
-    Runnable anonymous = new Runnable() {
-        @Override
+    // violation below ''{' at column 41 should have line break after'
+    Runnable anonymous = new Runnable() { @Override
         public void run() {
         }
     };

--- a/src/test/resources/com/puppycrawl/tools/checkstyle/checks/blocks/leftcurly/InputLeftCurlyTestTypesEol.java
+++ b/src/test/resources/com/puppycrawl/tools/checkstyle/checks/blocks/leftcurly/InputLeftCurlyTestTypesEol.java
@@ -1,0 +1,42 @@
+/*
+LeftCurly
+option = (default)eol
+ignoreEnums = false
+tokens = (default)ANNOTATION_DEF, CLASS_DEF, CTOR_DEF, ENUM_CONSTANT_DEF, ENUM_DEF, \
+         INTERFACE_DEF, LAMBDA, LITERAL_CASE, LITERAL_CATCH, LITERAL_DEFAULT, \
+         LITERAL_DO, LITERAL_ELSE, LITERAL_FINALLY, LITERAL_FOR, LITERAL_IF, \
+         LITERAL_SWITCH, LITERAL_SYNCHRONIZED, LITERAL_TRY, LITERAL_WHILE, METHOD_DEF, \
+         OBJBLOCK, STATIC_INIT, RECORD_DEF, COMPACT_CTOR_DEF, SWITCH_RULE
+
+
+*/
+
+package com.puppycrawl.tools.checkstyle.checks.blocks.leftcurly;
+
+public class InputLeftCurlyTestTypesEol {}
+
+// violation below ''{' at column 11 should have line break after'
+class Foo { int a;
+}
+
+// violation below ''{' at column 15 should have line break after'
+interface Bar { int b = 1;
+}
+
+// violation below ''{' at column 17 should have line break after'
+record FooBar() { static int c;
+}
+
+// violation below ''{' at column 12 should have line break after'
+enum  Bazz { VALUE,
+    VALUE2
+}
+
+// violation below ''{' at column 17 should have line break after'
+@interface Foo2 { int d = 1;
+}
+
+// violation below ''{' at column 12 should have line break after'
+class Bar2 {{
+    int a = 1;
+}}

